spring-cloud-gateway-core/src/main/java/org/springframework/cloud/gateway/route/CompositeRouteDefinitionLocator.java

```diff
@@ -51,8 +50,7 @@ public CompositeRouteDefinitionLocator(Flux<RouteDefinitionLocator> delegates,
 	public Flux<RouteDefinition> getRouteDefinitions() {
 		return this.delegates.flatMap(RouteDefinitionLocator::getRouteDefinitions)
 				.flatMap(routeDefinition -> Mono.justOrEmpty(routeDefinition.getId())
-						.defaultIfEmpty(idGenerator.generateId().toString())
-						.publishOn(Schedulers.elastic()).map(id -> {
```

Comment:
The trouble is generating an ID is blocking and therefor cannot be on the main thread.

@simonbasle or @bsideup ideas on how we might go about this?

Reply:
I would replace this `flatMap` with `.publishOn(Schedulers.boundedElastic()).map()` and remove `Mono.justOrEmpty`, it can be just `if`

Reply:
@bsideup does that move the running of the id generator to the other thread or does everything stay on that? The problem is that errors aren't bubbling up to the main thread now.

Reply:
what is "main thread" here? The one that callls `getRouteDefinitions()`?